### PR TITLE
fix(data): render and key the preprocess cache on data.chat_template

### LINKED ISSUE #1067: data.chat_template is not in the preprocess cache key, so two templates share one cache
Split out of #876, where @Srinivasan8888 found it while building the leading-BOS fix. The credit for the finding is his.

## What

`data.chat_template` is not part of the key that identifies a `soup data preprocess` cache. Two runs that differ only in their chat template therefore resolve to the same cache entry, and the second one silently trains on tokens rendered by the first one's template.

That is the same disease as #876 — a cache that does not reflect what produced it — but it is a different fix with a different blast radius, which is why it is here rather than widened into that issue.

## Why it is worse than it sounds

A wrong chat template does not crash and does not look wrong in the loss curve. It shifts where the assistant spans begin, so a run can train on the prompt, or mask the wrong turns, and complete normally. The v0.73.3 precedent is the shape to keep in mind: a label mask built from the wrong thing produced **0 trained tokens** with an ordinary-looking loss and no exception.

## The part that needs a decision, not just a patch

Adding a field to the key is one line. What to do about caches that already exist on disk is not:

1. **Invalidate** — the new key simply does not match, so every existing cache is rebuilt on next use. Correct, and costs users a re-preprocess they did not ask for.
2. **Migrate** — read the old key, check whether the template matches, adopt it if so. More code, and it has to guess what template produced a cache that never recorded one.
3. **Refuse** — a cache whose key predates this change is rejected with a message saying to re-run `soup data preprocess`. Loud, and does not pretend to know what it cannot know.

My lean is 3, on the grounds that a cache with no recorded template is a cache we cannot reason about, and silently rebuilding it (1) hides a correctness event behind a delay. But this is the acceptance decision the issue turns on, so it is worth arguing before anyone writes code.

## Acceptance criteria

- [ ] `data.chat_template` participates in the cache key, verified by two configs that differ only in that field resolving to **different** entries.
- [ ] A test that fails if the field is removed from the key derivation — not a test that the key changed, a test that the two configs no longer collide.
- [ ] A decision from the three above, implemented, with the user-visible behaviour documented in `docs/data.md`.
- [ ] A pre-existing cache (one written before this change) hits the chosen path, and the test proves which.
- [ ] Audit the rest of the key: if `chat_template` was missing, name every other field that reaches the tokenised output and confirm each is either in the key or explicitly does not affect it. **This is the criterion that keeps the issue from recurring with a different field name.**

## Not in scope

The leading-BOS disagreement between the cache and the live path stays on #876.

Related: #876 (the other half), #1054 (the preprocess cache has no labels, so a `pre_tokenized` SFT run trains on the prompt), #781 (inference encoding the prompt the way training encoded it).

### PR DESCRIPTION
**Pull request by Blizzeq**
## What does this PR do?

Fixes #1067. One thing the issue does not mention: `soup data preprocess` never applied `data.chat_template` at all. It rendered every row with the tokenizer's shipped template, so the cache was wrong for any config with an override, not only when two configs collided. Preprocess now renders with the resolved template and hashes it into the key.

I went with option 3 from the issue. The schema token moves to `v4`, so a cache written before this fails the existing `cache hash mismatch ... re-run soup data preprocess` check instead of being guessed at. The training gate also uses the template from the `pre_tokenized` config, because training saves the tokenizer with that template. A cache rendered with a different one is refused.

Audit of the rest of the key: `base`, `max_length`, the source format and `train` plus `interleave` are in it. Three inputs still change the cached rows without being in the key, and I left them out because each needs its own decision on the training side:
- `val_split` changes which rows land in the train split.
- `replay`, `replay_ratio` and `replay_seed` mix extra rows in.
- `task` picks raw text (`pretrain`) or the chat path.

`tests/test_issue1067_chat_template_cache_key.py` runs the real preprocess and gate. Four of its six tests fail on `main`, and the other two are controls. The full suite passes locally except `test_worker_verification_survives_a_real_venv_launcher`, which fails the same way on a clean `main` here.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactor / cleanup
- [ ] Documentation
- [ ] Tests

## Checklist

- [x] `ruff check src/soup_cli/ scripts/ tests/` passes
- [x] `pytest tests/ -v` passes
- [x] Updated relevant docs (`README.md` and the matching page under `docs/`) if needed
- [x] Added a changelog fragment, or this change has no user-visible impact

### CODE REVIEW

**Comment by Blizzeq** [comment]
Thanks for sorting this out, and sorry for the collision. I should have commented on #1067 when I opened this. I'll do that from now on.

On the pre-existing caches I agree with option 3, and that is what this PR does. The schema token goes from `v3` to `v4`, so any cache written before the change fails the existing `cache hash mismatch ... re-run soup data preprocess` check. Nothing is rebuilt silently or guessed at. `test_a_cache_written_before_the_fix_is_refused` covers it.

CI is waiting for approval since this is my first PR here.

**Comment by Srinivasan8888** [comment]
@Blizzeq — heads-up on a collision between this and **#1096** (the #876 BOS half),
not a review ask.

I built the same `data.chat_template` render fix while measuring #876, found this
PR already had it, and reverted mine. Yours is older and assigned; nothing of mine
touches that code now.

What is left to coordinate is one line. Both PRs bump
`_PREPROCESS_TOKENIZE_SCHEMA` from `v3` to `v4`, and both change what the cache
stores. Whichever merges second has to go to `v5`, or a cache built between the
two merges looks current to the second change and is reused when it should be
rebuilt. I have said on #1096 that I am happy to be the one who rebases.

The other thing you may find useful: the two fixes need each other. Measured on
`unsloth/Meta-Llama-3.1-8B-Instruct` with `chat_template: llama3`, 9 chat rows,
`train_on_responses_only: false` — row 0, same shape on all 9:

| tree | live len | cache len | live leading BOS | cache leading BOS | byte-identical |
|---|---|---|---|---|---|
| `main` | 19 | 45 | 0 | 1 | no |
| #1096 alone | 19 | 45 | 0 | 1 | no |
| **this PR alone** | 19 | 20 | 0 | 1 | **no** |
| this PR + #1096 | 19 | 19 | 0 | 0 | yes |

The 20-vs-19 row is this PR on its own: the body is right after your change, and
the cache still carries one leading BOS the live path does not, because it
tokenises with `add_special_tokens=True` and the Llama post-processor prepends
one. #1096 removes exactly that. The probe I used is in
https://github.com/MakazhanAlpamys/Soup/pull/1096#issuecomment-5744142173 if you
want to reproduce it against your branch.

**Comment by Blizzeq** [comment]
@Srinivasan8888 thanks, that table is useful. Agreed on the schema token: whichever merges second goes to `v5`. Your offer to rebase #1096 works for me, and if this one lands second I'll bump it here instead.

The leading BOS is out of scope for this PR and stays with #876, as the issue says. So the table matches what I'd expect: this PR fixes the body, and byte-identical rows need both.

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Welcome, and this is a strong first PR — the code is right, both halves of the fix are genuinely pinned,
and you widened the scope in the one direction that was necessary and said so in the first paragraph
rather than smuggling it. Two things to fix, one of them mechanical, and then this merges.

## What I verified, by execution

```
mutation: drop the chat_template segment from the hash blob  -> 2 failed   KILLED
mutation: keep the key but revert the render (`if False:`)   -> 1 failed   KILLED
mutation: the training gate passes chat_template=None       -> 1 failed   KILLED
mutation: stop recording chat_template in metadata.json     -> 1 failed   KILLED
38 passed on your file + the 791/785/1038 suites; 332 passed merged with main; ruff clean.
On main your file gives 4 failed / 2 passed, and the two passers are your declared controls.
```

Your decision to render with the template as well as key on it is correct, and M1/M2 above are why I can
say so rather than guess: a key that changed without the render changing would have written two cache
entries that both held the wrong tokens.

I tested the pre-existing-cache policy against a cache **actually written by `main`'s code**, not a
hand-edited `metadata.json`: `main`'s preprocess wrote `.soup-tokenized/71dc713da3c8cec4` with no
`chat_template` key at all; `main`'s loader accepts it; your gate refuses it with
`pre_tokenized cache hash mismatch: … re-run 'soup data preprocess'`. Option 3, implemented.

And the collision itself, on ids rather than counts, three real `soup data preprocess` runs in one
directory:

```
main:     71dc713da3c8cec4 / 71dc713da3c8cec4 / 71dc713da3c8cec4   ids identical all three times
your PR:  5c6eb2046ecf71d9 / e772b84112c24458 / 67a8baddbbcf0a20   ids all different
```

## Blocking, and it is one edit in a test

`tests/test_issue791_preprocess_cache_eos.py:343`. You changed

```python
assert current == _blob_key("v3"), "current schema is v3 (#791)"
```

to a `!=`. Because your fix appends a **sixth** field to the hash blob, that helper now builds a
five-field blob — so all three `!=` assertions in that test are **vacuously true**: no value of
`_PREPROCESS_TOKENIZE_SCHEMA` can make a six-field blob equal a five-field one. Measured: forcing the
token to `"v2"` (a state in which a pre-#785 cache would be reused) gives **18 passed**.

That line was the repo's only *positive* pin on the schema token, so as it stands the constant can
regress silently. The fix is in the same test: add the sixth segment to its `_blob_key` helper (the empty
template) and restore `assert current == _blob_key("v4")`, keeping the `!=` lines for v3 and v2. #1096
does exactly this in its own file, so the pattern is already in the tree.

(Worth knowing: the token bump is *not* load-bearing today — with it reverted, your refusal test still
passes, because the extra blob field alone invalidates every old key. Belt and braces is good; a
constant nothing pins is not.)

## Also needed at merge: the credit line

`changelog.d/0.75.0/1082.fixed.md` reads `(#1067 in #1082)`. 99 of the 114 fragments in that directory
carry `by @<login>` — please make it `(#1067 by @Blizzeq in #1082)`. Your name on your work is the only
currency this project pays in, and I would rather you put it there than have me add it.

## Two smaller things, not blocking

1. **Your criterion-5 audit found three more holes and nothing tracks them.** `val_split`, the
   `replay*` trio and `task` all change the cached rows and are absent from the key — I confirmed that by
   reading `preprocess_dataset_key_input`. You were right to keep them out of this diff; they need their
   own issue before #1067 closes, and I will file it with your credit unless you would rather take it.
2. **The refusal message cannot say "your cache predates the template change."** A user with a v3 cache
   gets the same text as one who changed `max_length`. The field is already recorded, so one clause when
   `metadata.json` has no `chat_template` key would make it self-explaining.

## Ordering, so you are not surprised

**This PR merges first**, then #1096 (@Srinivasan8888's BOS half) rebases on top and takes the schema
token to `v5` — he offered exactly that on #876 and #1096. Your PR is the older one, the issue was
assigned on PR age, and #1096 depends on yours to demonstrate its own headline case, so the order is not
a judgement about quality. Nothing else in the two overlaps: `commands/data.py` auto-merges cleanly, and I
checked the merged result is semantically right as well as textually.

Push those two and I will merge.

**Comment by MakazhanAlpamys** [comment]
**Addendum to my review, and a correction to what I told @Srinivasan8888 on #1096: there are THREE PRs on
this constant, not two.** #1120 (@AaronProbha18, the `pre_tokenized` labels fix for #1054) also bumps
`_PREPROCESS_TOKENIZE_SCHEMA`, and it was that review that surfaced the hazard.

```
main     v3   blob: ...{max_length}\x1f{format_name}
#1120    v4   blob: ... + \x1f{mask_mode}          <- adds a key input
#1082    v4   blob: ... + \x1f{chat_template}      <- adds a key input
#1096    v4   blob: ... unchanged                  <- adds NONE; the token is its ONLY invalidation
```

All three land on **exactly `v4`**. #1120 and #1082 each add a key input, so they invalidate old caches
whatever the token says. **#1096 adds none**, so if either of the others merges first and the tree is
already at `v4`, then #1096 merges second still reading `v4` — and a cache written between the two merges is
**silently reused** by the post-#1096 tree even though #1096 changes the row's leading BOS. That is the
family of defect all three of you are fixing, reintroduced by merge order alone.

**The rule, posted on all three threads: second to merge goes to `v5`, third to `v6`.** And
`tests/test_issue791_preprocess_cache_eos.py::TestCacheKey` — which all three of you edit, and which asserts
`current == _blob_key("v4")` — is re-pointed in the same commit, along with each PR's own `TestCacheKey`
helper if it hard-codes the blob shape (three of them now do, which is why this keeps surfacing as a merge
surprise).

Merge order between the three is unchanged where I have ruled it: **#1082 first**, then #1096. #1120 has two
blocking findings of its own, so I expect it third — but if it becomes ready first, the same arithmetic
applies with the numbers shifted, and I will say so explicitly rather than leaving you to infer it.

Nothing about your own diff changes because of this note.

**Comment by MakazhanAlpamys** [comment]
Heads-up that affects this PR mechanically, not its content: the changelog fragment baseline is moving from `changelog.d/0.75.0/` to `changelog.d/0.75.1/`.

Why: v0.75.1 shipped to PyPI while `main`'s CHANGELOG still stopped at 0.75.0, and four pending fragments described fixes that had already gone out in it — left alone, the next release would have announced them as new. Adding the 0.75.1 heading is the fix, and `validate_fragments()` requires the fragment directory to be named for the newest heading, so every pending fragment moves with it.

What it means for you: once that lands, this PR's changelog check goes red with

> `changelog.d/0.75.0/NNNN.category.md: fragment baseline is 0.75.0, but the latest release is 0.75.1; rebase and move the fragment before merging`

and the fix is one line on your next push:

```
git mv changelog.d/0.75.0/NNNN.category.md changelog.d/0.75.1/
```

Nothing about your diff is wrong and no review finding changes. I am telling you rather than letting a red X be the first you hear of it.

**Comment by Blizzeq** [comment]
Both are in the latest commit.

`test_issue791`'s `_blob_key` now takes the template segment, and the positive pin is back as `current == _blob_key("v4", chat_template="")`, with the `!=` lines for v3, v2 and the pre-schema blob kept. Forcing the token to `v2`, `v3` or `v5` now fails it. The fragment reads `(#1067 by @Blizzeq in #1082)`.

I also took the second smaller point. When `metadata.json` has no `chat_template` key, the refusal now says the cache predates template keying, and a control test checks that a mismatch on a current cache does not say so. Both mutations of that condition fail a test.

**Comment by MakazhanAlpamys** [comment]
Merged as `3bcd2f04` — your first merge here. Welcome, and thank you: this closes a silent-wrong-data defect, which is the worst kind, because nothing fails and the run still finishes.

Verified before merging, on head `a09800c2`:

- **The defect reproduced on `main` first**, so we know what was fixed: a config with `data.chat_template` set produced a cache holding the **shipped** template's markup, and an unregistered template name was accepted silently with exit 0.
- **The blocker is closed and the pin is real.** Forcing `_PREPROCESS_TOKENIZE_SCHEMA` to `v2` — the exact state that left **18 tests green** last round — now fails, as do `v3` and `v5`. The assertion that fires is the restored positive pin, with the two keys printed side by side.
- **Nine mutations, nine kills, and each half has its own killer**: dropping the template segment from the blob fails 4, removing the render fails 1, dropping the training-side gate fails 1, dropping the `metadata.json` field fails 2, and both directions of the new refusal message fail 1 each. On `main`, five of your six new tests fail and the declared control passes.
- **Your "cache written before the fix" fixture is a real key, not a stand-in** — I ran `main`'s own `make_preprocess_cache_key` on that config and got exactly the key your test writes, byte for byte. That detail is why the refusal test means what it says.
- 13 passed on the two files, 354 across every consumer of `make_preprocess_cache_key` / `pre_tokenized`, 375 on the tree merged with current `main`, ruff clean, **13/13 required checks green**. (The 24 failures I saw in the wide `chat_template` sweep are #776, the NF4 divergence on this Blackwell card — identical IDs on a clean `main`, so zero regressions from you.)

**Two things I want to name.** First, the scope widening — rendering, not just the key — was necessary and you declared it in the first paragraph rather than letting me find it; without it the fix would have written two cache entries that both held the wrong tokens. Second, when the criterion-5 audit turned up three more fields the key omits, you filed them as #1127 instead of quietly expanding this PR. Both are exactly right.

Two follow-ups are tracked rather than folded in, so you are not owed anything here:
- **#1127** — `val_split`, the `replay*` fields and `task` are still absent from the key.
- An **unregistered** template name now surfaces as a bare `KeyError` from `resolve_chat_template` at `commands/data.py:2409`. It is loud and it lists the valid names, and failing loudly beats caching the wrong tokens — but it is a traceback where the rest of `preprocess_dataset` uses `console.print` + `Exit(1)`. The same bare `KeyError` already escapes the live training path, so it predates you; if you want it, it is a one-line `try/except` at two call sites and I would review it happily.

Small thing for next time: `test_issue791_preprocess_cache_eos.py:344/346` still build **five**-field blobs, so those `!=` assertions are vacuously true and their messages claim more than they test. Harmless now that the `==` pin exists — but whoever re-points that block to v5 should not mistake them for load-bearing.
